<a href="https://colab.research.google.com/github/007-athul/DSA-case-study1/blob/main/data_acquisition_casestudy.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

Part A – Data Extraction

Task 1
Read the Population dataset from the CSV file.
Hint:
● Use the pandas library.
● Display the first five rows.
● Check the shape and data types.

In [79]:
import pandas as pd

In [80]:
pop_data=pd.read_csv('population.csv')
pop_data.head()

,country,population
0,United States,331002651
1,China,1411778724
2,India,1380004385
3,Japan,125836021
4,Germany,83240525


In [81]:
pop_data.shape

(25, 2)

In [82]:
pop_data.dtypes

,0
country,object
population,int64


Task 2
Read the GDP dataset from the Excel file.
Hint:
● Use the appropriate pandas function.
● Display column names and data types.

In [83]:
gdp_data=pd.read_excel('gdp.xlsx')
gdp_data.head()

,Country,GDP
0,United States,30500000
1,China,19230000
2,India,4187000
3,Japan,4180000
4,Germany,4744000


In [84]:
gdp_data.dtypes

,0
Country,object
GDP,int64


Task 3
Read the Internet Users dataset from the JSON file.
Hint:
● Use either the json module or pandas.
● Convert the data into a DataFrame.

In [85]:
json_data=pd.read_json('internet_users.json')
json_data.head()

,country,internet_users
0,United States,312000000
1,China,1080000000
2,India,850000000
3,Japan,118000000
4,Germany,79000000


In [86]:
json_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 25 entries, 0 to 24
Data columns (total 2 columns):
 #   Column          Non-Null Count  Dtype 
---  ------          --------------  ----- 
 0   country         25 non-null     object
 1   internet_users  25 non-null     int64 
dtypes: int64(1), object(1)
memory usage: 532.0+ bytes


Task 4
Read the Literacy Rate dataset from the XML file.
Hint:
● Use xml.etree.ElementTree or pandas.read_xml().
● Extract all required fields into a DataFrame.

In [87]:
lit_data=pd.read_xml('literacy_rate.xml')
lit_data.head()

,name,literacy_rate
0,United States,99.0
1,China,97.2
2,India,77.7
3,Japan,99.0
4,Germany,99.0


In [88]:
lit_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 25 entries, 0 to 24
Data columns (total 2 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   name           25 non-null     object 
 1   literacy_rate  25 non-null     float64
dtypes: float64(1), object(1)
memory usage: 532.0+ bytes


Part B – Data Understanding
Task 5
Explore each dataset.
Find the following:
● Number of rows
● Number of columns
● Column names
● Data types
● Missing values
● Duplicate records

In [89]:
#population dataset
print("Population Dataset\n")
rows, cols = pop_data.shape
print(f"Number of rows: {rows}")
print(f"Number of columns: {cols}")
print("\nColumn names:")
print(pop_data.columns.tolist())

print("\nData types:")
print(pop_data.dtypes)

print("\nMissing values per column:")
print(pop_data.isnull().sum())

duplicate_count = pop_data.duplicated().sum()
print(f"\nTotal duplicate records: {duplicate_count}")

Population Dataset

Number of rows: 25
Number of columns: 2

Column names:
['country', 'population']

Data types:
country       object
population     int64
dtype: object

Missing values per column:
country       0
population    0
dtype: int64

Total duplicate records: 0


In [90]:
#Literacy Rate

print("Literacy Rate\n")
rows, cols = lit_data.shape
print(f"Number of rows: {rows}")
print(f"Number of columns: {cols}")
print("\nColumn names:")
print(lit_data.columns.tolist())

print("\nData types:")
print(lit_data.dtypes)

print("\nMissing values per column:")
print(lit_data.isnull().sum())

duplicate_count = lit_data.duplicated().sum()
print(f"\nTotal duplicate records: {duplicate_count}")

Literacy Rate

Number of rows: 25
Number of columns: 2

Column names:
['name', 'literacy_rate']

Data types:
name              object
literacy_rate    float64
dtype: object

Missing values per column:
name             0
literacy_rate    0
dtype: int64

Total duplicate records: 0


In [91]:
#Internet Users dataset
print("Internet Users Dataset\n")
rows, cols = json_data.shape
print(f"Number of rows: {rows}")
print(f"Number of columns: {cols}")
print("\nColumn names:")
print(json_data.columns.tolist())

print("\nData types:")
print(json_data.dtypes)

print("\nMissing values per column:")
print(json_data.isnull().sum())

duplicate_count = json_data.duplicated().sum()
print(f"\nTotal duplicate records: {duplicate_count}")

Internet Users Dataset

Number of rows: 25
Number of columns: 2

Column names:
['country', 'internet_users']

Data types:
country           object
internet_users     int64
dtype: object

Missing values per column:
country           0
internet_users    0
dtype: int64

Total duplicate records: 0


In [92]:
#gdp dataset

print("GDP Dataset\n")
rows, cols = gdp_data.shape
print(f"Number of rows: {rows}")
print(f"Number of columns: {cols}")
print("\nColumn names:")
print(gdp_data.columns.tolist())

print("\nData types:")
print(gdp_data.dtypes)

print("\nMissing values per column:")
print(gdp_data.isnull().sum())

duplicate_count = gdp_data.duplicated().sum()
print(f"\nTotal duplicate records: {duplicate_count}")

GDP Dataset

Number of rows: 25
Number of columns: 2

Column names:
['Country', 'GDP']

Data types:
Country    object
GDP         int64
dtype: object

Missing values per column:
Country    0
GDP        0
dtype: int64

Total duplicate records: 0


Task 6
Rename columns wherever necessary so that all datasets have a common column
name for merging.
Hint:
● Example: Country → country

In [93]:
gdp_data.rename(columns={'Country': 'country'},inplace =True)
lit_data.rename(columns={'name':'country'},inplace=True)

print("GDP columns:", gdp_data.columns.tolist())
print("Literacy Rate columns:", lit_data.columns.tolist())
print("Population columns:", pop_data.columns.tolist())
print("Internet Users columns:", json_data.columns.tolist())

GDP columns: ['country', 'GDP']
Literacy Rate columns: ['country', 'literacy_rate']
Population columns: ['country', 'population']
Internet Users columns: ['country', 'internet_users']


Task 7
Convert the required columns into appropriate numeric data types.

In [94]:

pop_data['population'] = pop_data['population'].astype('int64')
gdp_data['GDP'] = gdp_data['GDP'].astype('int64')
json_data['internet_users'] = json_data['internet_users'].astype('int64')
lit_data['literacy_rate'] = lit_data['literacy_rate'].astype('float64')

print(pop_data.dtypes)
print(gdp_data.dtypes)
print(json_data.dtypes)
print(lit_data.dtypes)

country       object
population     int64
dtype: object
country    object
GDP         int64
dtype: object
country           object
internet_users     int64
dtype: object
country           object
literacy_rate    float64
dtype: object


Task 8
Merge all four datasets into one DataFrame.
Hint:
● Use the common column (country)
● Choose an appropriate join method.

In [95]:
merged_data= pd.merge(pop_data,gdp_data, on='country', how='inner')
merged_data= pd.merge(merged_data,json_data, on='country', how='inner')

final_data= pd.merge(merged_data,lit_data, on='country', how='inner')
final_data.head()

,country,population,GDP,internet_users,literacy_rate
0,United States,331002651,30500000,312000000,99.0
1,China,1411778724,19230000,1080000000,97.2
2,India,1380004385,4187000,850000000,77.7
3,Japan,125836021,4180000,118000000,99.0
4,Germany,83240525,4744000,79000000,99.0


Task 9
Create the following calculated columns:
1. Internet Penetration (%)
2. GDP Per Capita

Hint:
Round the calculated values to two decimal places.

In [96]:
final_data['Internet Penetration (%)']=((final_data['internet_users']/ final_data['population'])*100).round(2)
final_data['GDP Per Capita ']=(final_data['GDP']/final_data['population'])
final_data.head()


,country,population,GDP,internet_users,literacy_rate,Internet Penetration (%),GDP Per Capita
0,United States,331002651,30500000,312000000,99.0,94.26,0.092144
1,China,1411778724,19230000,1080000000,97.2,76.50,0.013621
2,India,1380004385,4187000,850000000,77.7,61.59,0.003034
3,Japan,125836021,4180000,118000000,99.0,93.77,0.033218
4,Germany,83240525,4744000,79000000,99.0,94.91,0.056991


Part D – Load Data
Task 10
Create a SQLite database. Store the final merged dataset into a table named
‘country_statistics’
Hint:
● Use the sqlite3 library.
● Load the DataFrame using pandas.


In [97]:
import sqlite3 as sq

conn= sq.connect('world_data.db')
final_data.to_sql('country_statistics', conn, if_exists='replace', index=False)
conn.close()
print("Database Created sucessfully")

Database Created sucessfully


Task 11
Verify that the data has been loaded successfully.
Hint:
Retrieve the first five records from the SQLite table.

In [98]:
conn=sq.connect('world_data.db')
query="SELECT * FROM country_statistics LIMIT 5;"
verify_data=pd.read_sql_query(query,conn)
conn.close()
verify_data



,country,population,GDP,internet_users,literacy_rate,Internet Penetration (%),GDP Per Capita
0,United States,331002651,30500000,312000000,99.0,94.26,0.092144
1,China,1411778724,19230000,1080000000,97.2,76.50,0.013621
2,India,1380004385,4187000,850000000,77.7,61.59,0.003034
3,Japan,125836021,4180000,118000000,99.0,93.77,0.033218
4,Germany,83240525,4744000,79000000,99.0,94.91,0.056991


Part E – Data Analysis Using Pandas
Answer the following questions using pandas.

In [99]:
#1 List the Top 5 countries by GDP.
top5_gdp= final_data.sort_values(by='GDP',ascending=False).head()[['country','GDP']]
top5_gdp

,country,GDP
0,United States,30500000
1,China,19230000
4,Germany,4744000
2,India,4187000
3,Japan,4180000


In [100]:
#2 List the Top 5 countries by population.
top5_pop= final_data.sort_values(by='population',ascending=False).head()[['country','population']]
top5_pop

,country,population
1,China,1411778724
2,India,1380004385
0,United States,331002651
15,Indonesia,273523615
10,Brazil,212559417


In [101]:
#3 Calculate the average literacy rate.
avg_literacy=final_data['literacy_rate'].mean()
print(avg_literacy)

97.248


In [102]:
#4 List all countries having a literacy rate greater than 90%.
high_literacy=final_data[final_data['literacy_rate'] >90][['country','literacy_rate']]
high_literacy

,country,literacy_rate
0,United States,99.0
1,China,97.2
3,Japan,99.0
4,Germany,99.0
5,United Kingdom,99.0
6,France,99.0
7,Italy,99.0
8,Canada,99.0
9,Australia,99.0
10,Brazil,93.2


In [103]:
#5 List all countries having an internet penetration greater than 70%.
high_internet_pen=final_data[final_data['Internet Penetration (%)'] >70][['country','Internet Penetration (%)']]
high_internet_pen

,country,Internet Penetration (%)
0,United States,94.26
1,China,76.50
3,Japan,93.77
4,Germany,94.91
5,United Kingdom,97.22
6,France,94.98
7,Italy,92.62
8,Canada,94.72
9,Australia,93.43
10,Brazil,85.15


In [104]:
#6 List the Top 5 countries by GDP per capita
top5_gdp_per_capita=final_data.sort_values(by='GDP Per Capita ',ascending=False).head()
top5_gdp_per_capita

,country,population,GDP,internet_users,literacy_rate,Internet Penetration (%),GDP Per Capita
21,Switzerland,8654622,940000,8500000,99.0,98.21,0.108612
23,Norway,5421241,580000,5300000,100.0,97.76,0.106987
0,United States,331002651,30500000,312000000,99.0,94.26,0.092144
24,Singapore,5850342,530000,5600000,97.5,95.72,0.090593
9,Australia,25687041,1880000,24000000,99.0,93.43,0.073189


In [105]:
#7 Calculate the total population represented in the dataset.
tot_pop=final_data['population'].sum()
print(tot_pop)

4720039725


In [106]:
#8 Calculate the average internet penetration across all countries.
avg_pen=final_data['Internet Penetration (%)'].mean()
print(avg_pen)

89.6596


In [107]:
#9 Identify the country having the highest number of internet users.
highest_int_users=final_data.sort_values(by='internet_users',ascending=False).head(1)[['country','internet_users']]
highest_int_users

,country,internet_users
1,China,1080000000


In [108]:
#10 Find the correlation between Literacy Rate and Internet Penetration.
correlation= final_data['literacy_rate'].corr(final_data['Internet Penetration (%)'])
print(correlation)

0.77279756813985


Part F – Data Analysis Using SQL
Answer the following questions using SQLite queries.

In [109]:
#11 Display the Top 5 countries by GDP.
conn=sq.connect('world_data.db')
query="SELECT country,GDP FROM country_statistics ORDER BY GDP DESC LIMIT 5;"
top5_gdp=pd.read_sql_query(query,conn)
top5_gdp

,country,GDP
0,United States,30500000
1,China,19230000
2,Germany,4744000
3,India,4187000
4,Japan,4180000


In [110]:
#12 Find the average literacy rate.
query= "SELECT AVG(literacy_rate) FROM country_statistics;"
avg_literacy_rate=pd.read_sql_query(query,conn)
avg_literacy_rate


,AVG(literacy_rate)
0,97.248


In [111]:
#13 Display all countries having literacy rate above 90%.
query="SELECT country,literacy_rate FROM country_statistics WHERE literacy_rate>90;"
lit_rate_above90=pd.read_sql_query(query,conn)
lit_rate_above90

,country,literacy_rate
0,United States,99.0
1,China,97.2
2,Japan,99.0
3,Germany,99.0
4,United Kingdom,99.0
5,France,99.0
6,Italy,99.0
7,Canada,99.0
8,Australia,99.0
9,Brazil,93.2


In [112]:
#14Find the Top 5 countries with the highest internet penetration.
query="SELECT country,[INTERNET PENETRATION (%)] from country_statistics ORDER BY [INTERNET PENETRATION (%)] DESC LIMIT 5;"
top5_intpen=pd.read_sql_query(query,conn)
top5_intpen

,country,Internet Penetration (%)
0,Switzerland,98.21
1,Norway,97.76
2,Saudi Arabia,97.66
3,Sweden,97.55
4,Netherlands,97.47


In [113]:
#15 Calculate the total population.
query="SELECT SUM(population) AS total_population FROM country_statistics;"
tot_pop=pd.read_sql_query(query,conn)
tot_pop

,total_population
0,4720039725


In [114]:
#16 Find the country with the highest GDP per capita.
query_16 = "SELECT country, [GDP Per Capita ] FROM country_statistics ORDER BY [GDP Per Capita ] DESC LIMIT 1;"
highest_gdp_capita = pd.read_sql_query(query_16, conn)
highest_gdp_capita

,country,GDP Per Capita
0,Switzerland,0.108612


In [115]:
#17 Export the final merged dataset as a CSV file.
final_data.to_csv('final_merged_data.csv', index=False)
print("Dataset exported successfully as 'final_merged_data.csv'")

Dataset exported successfully as 'final_merged_data.csv'


In [116]:
#18 Sort the dataset in descending order of GDP per Capita and display the Top 10 countries.
top_10_gdp_capita = final_data.sort_values(by='GDP Per Capita ', ascending=False).head(10)
top_10_gdp_capita

,country,population,GDP,internet_users,literacy_rate,Internet Penetration (%),GDP Per Capita
21,Switzerland,8654622,940000,8500000,99.0,98.21,0.108612
23,Norway,5421241,580000,5300000,100.0,97.76,0.106987
0,United States,331002651,30500000,312000000,99.0,94.26,0.092144
24,Singapore,5850342,530000,5600000,97.5,95.72,0.090593
9,Australia,25687041,1880000,24000000,99.0,93.43,0.073189
20,Netherlands,17441139,1150000,17000000,99.0,97.47,0.065936
22,Sweden,10353442,650000,10100000,99.0,97.55,0.062781
8,Canada,38005238,2250000,36000000,99.0,94.72,0.059202
4,Germany,83240525,4744000,79000000,99.0,94.91,0.056991
5,United Kingdom,67886011,3839000,66000000,99.0,97.22,0.056551


In [117]:
conn.close()

Data Extraction & Sources: Extracted matching 25-row datasets from four distinct file formats—Population (.csv), GDP (.xlsx), Internet Users (.json), and Literacy Rate (.xml)—using tailored pandas functions (read_csv, read_excel, read_json, read_xml).

Data Cleaning: Standardized structural differences by renaming inconsistent tracking identifiers (like Country and name) to a unified lowercase column named 'country', and explicitly cast all quantitative fields to appropriate numeric data types (int64 and float64).

Data Transformation: Performed an inner-join merge across all four cleaned dataframes based on the matching 'country' field, and engineered two new calculated features: Internet Penetration (%) and GDP Per Capita, with values properly rounded.

Database Loading: Opened a connection using sqlite3, structured the fully transformed dataset, and written it directly into an SQLite data infrastructure table named 'country_statistics' using the to_sql() method.

Analysis Queries: Conducted multi-angled programmatic analysis using both pandas functions and structured SQLite analytical queries (ORDER BY, LIMIT, aggregations like SUM and AVG) to extract comparative insights across data points.

Key Analytical Insights: Identified China as having the highest total number of internet users, while Switzerland led the dataset in GDP per Capita. Additionally, the analysis revealed an average global literacy rate of 97.25% and a strong positive correlation (0.77) between a nation's literacy rate and its internet penetration percentage.